posts/2019-07-04-Rust-1.36.0.md

```diff
@@ -0,0 +1,219 @@
+a new `HashMap<K, V>` implementation, and [`--offline`] support in Cargo.
+Read on for a few highlights, or see the [detailed release notes][notes] for additional information.
+
+### The [`Future`] is here!
+
+[`Future`]: https://doc.rust-lang.org/std/future/trait.Future.html
+[pr-future]: https://github.com/rust-lang/rust/pull/59739
+
+In Rust 1.36.0 the long awaited [`Future`] trait has been [stabilized][pr-future]!
+
+With this stabilization, we hope to give important crates, libraries,
+and the ecosystem time to prepare for `async` / `.await`,
+which we'll tell you more about in the future.
+
+### The [`alloc`][alloc-crate] crate is stable
+
+[alloc-crate]: https://doc.rust-lang.org/alloc/index.html
+
+Before 1.36.0, the standard library consisted of the crates `std`, `core`, and `proc_macro`.
+The `core` crate provided core functionality such as `Iterator` and `Copy`
+and could be used in `#![no_std]` environments since it did not impose any requirements.
+Meanwhile, the `std` crate provided types like `Box<T>` and OS functionality
+but required a global allocator and other OS capabilities in return.
+
+Starting with Rust 1.36.0, the parts of `std` that depend on a global allocator, e.g. `Vec<T>`,
+are now available in the `alloc` crate. The `std` crate then re-exports these parts.
+While `#![no_std]` binaries using `alloc` still require nightly Rust,
+`#![no_std]` library crates can use the `alloc` crate in stable Rust.
+Meanwhile, normal binaries, without `#![no_std]`, can depend on such library crates.
+We hope this will facilitate the development of a `#![no_std]` compatible ecosystem of libraries
+prior to stabilizing support for `#![no_std]` binaries using `alloc`.
+
+If you are the maintainer of a library that only relies on some allocation primitives to function,
+consider making your library `#[no_std]` compatible by using the following at the top of your `lib.rs` file:
+
+```rust
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+```
+
+### [`MaybeUninit<T>`] instead of [`mem::uninitialized`]
+
+[`MaybeUninit<T>`]: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html
+[`mem::uninitialized`]: https://doc.rust-lang.org/std/mem/fn.uninitialized.html
+[gankro-blog]: https://gankro.github.io/blah/initialize-me-maybe/
+[pr-60445]: https://github.com/rust-lang/rust/pull/60445
+
+In previous releases of Rust, the [`mem::uninitialized`] function has allowed you to bypass Rust's
+initialization checks by pretending that you've initialized a value at type `T` without doing anything.
+One of the main uses of this function has been to lazily allocate arrays.
+
+However, [`mem::uninitialized`] is an incredibly dangerous operation that essentially
+cannot be used correctly as the Rust compiler assumes that values are properly initialized.
+For example, calling `mem::uninitialized::<bool>()` causes *instantaneous __undefined behavior__*
+as the random bits in memory could be something other than `0` (for `false`) and `1` (for `true`)
```

Comment:
Maybe saying "uninitialized bits" here is a bit better, to at least not *explicitly* perpetuate the myth about uninitialized memory being "just" uninitialized bits? The linked blog post then gives further details.

Reply:
Updated, please take a look.

Reply:
Much better, thanks!

The "could" here still sounds like there is any non-determinism involved, which in fact there is not. Do you think it would be confusing to replace it by "are"?

> For example, calling `mem::uninitialized::<bool>()` causes *instantaneous __undefined behavior__* as the uninitialized bits are neither `0` (for `false`) nor `1` (for `true`)  -- the only two allowed bit patterns for `bool`.

Reply:
I wonder whether we should be clearer about "abstract machine vs. low-level implementation perspective" by saying:

> For example, calling `mem::uninitialized::<bool>()` causes _instantaneous **undefined behavior**_ as, according to Rust's abstract machine, the uninitialized bits are neither `0` (for `false`) nor `1` (for `true`)  -- the only two allowed bit patterns for `bool`.

Reply:
Changed to "from Rust's point of view" + "are" per discussion on Zulip. Thanks!